disk-utils/cfdisk.c

```diff
@@ -2752,19 +2761,34 @@ int main(int argc, char *argv[])
 	} else
 		diskpath = argv[optind];
 
-	rc = fdisk_assign_device(cf->cxt, diskpath, 0);
-	if (rc == -EACCES)
-		rc = fdisk_assign_device(cf->cxt, diskpath, 1);
+        uid = getuid();
+        if (!read_only){
+	        rc = fdisk_assign_device(cf->cxt, diskpath, 0);
+
+                /*Forced try to open in read-only mode */
+	        if (rc == -EACCES){
+                        struct stat disk_perms;
+                        if (uid){
+                                if (!stat(diskpath, &disk_perms)){
+                                        if(!(disk_perms.st_mode & S_IROTH))
+                                                err(EXIT_FAILURE, _("reading of %s"), diskpath);
+                                }
+                                rc = fdisk_assign_device(cf->cxt, diskpath, 1);
+                        }
+                        else
+                                err(EXIT_FAILURE, _("writing to %s denied to root"), diskpath);
+                }
+                
+        } else {
+                rc = fdisk_assign_device(cf->cxt, diskpath, 1);
+        }
 	if (rc != 0)
 		err(EXIT_FAILURE, _("cannot open %s"), diskpath);
+	if (blkdev_lock(fdisk_get_devfd(cf->cxt), diskpath, lockmode) != 0)
+	        return EXIT_FAILURE;
 
-	if (!fdisk_is_readonly(cf->cxt)) {
-		if (blkdev_lock(fdisk_get_devfd(cf->cxt), diskpath, lockmode) != 0)
-			return EXIT_FAILURE;
-
-		cf->device_is_used = fdisk_device_is_used(cf->cxt);
-		fdisk_get_partitions(cf->cxt, &cf->original_layout);
-	}
+	cf->device_is_used = fdisk_device_is_used(cf->cxt);
+	fdisk_get_partitions(cf->cxt, &cf->original_layout);
 
```

Comment:
It's better to rely on fdisk_is_readonly(cf->cxt) than on read_only. It returns the final access mode used by the library (don't forget we open read-only on EACCES independently on read_only setting.